### PR TITLE
fix: add validation to prevent Enterprise systemwideroleassignment fr...

### PR DESCRIPTION
**Pull request by christopappas**
…om being saved with null enterprise_customer if has_access_to_all_contexts is not True

also, removing an old management command and all its tests (confirmed internally it is no longer needed)

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/models.py [hunk 1]
```diff
@@ -2849,6 +2849,25 @@ class SystemWideEnterpriseUserRoleAssignment(UserRoleAssignment):
...
+
+        if self.enterprise_customer is None and self.has_access_to_all_contexts() is not True:
+            message = (
+                "Enterprise customer must be set on SystemWideEnterpriseUserRoleAssignment "
+                f"instance {self.id} because has_access_to_all_contexts is not True."
+            )
+            raise ValidationError(message)
+
+    def save(self, *args, **kwargs):
+        self._validate_whether_enterprise_customer_needed()
```
Comment (macdiesel) [inline enterprise/models.py:2868]: I'm worried that this exception message is going to end up getting swallowed and logged in a strange way... Should I be concerned? Would it be better to just put the message to the logger in the `_validate_whether_enterprise_customer_needed` method? 

also this save method now also raises `ValidationError`.

**Review by macdiesel — APPROVED** [review]
One nit.

File: enterprise/models.py [hunk 2]
```diff
@@ -2849,6 +2849,25 @@ class SystemWideEnterpriseUserRoleAssignment(UserRoleAssignment):
...
 
+    def _validate_whether_enterprise_customer_needed(self):
+        """
+        Verify that if enterprise_customer is None, then has_access_to_all_contexts
+        must be true.
+
+        Raises ValidationError
+        """
+
+        if self.enterprise_customer is None and self.has_access_to_all_contexts() is not True:
```
Comment (long74100) [inline enterprise/models.py:2860]: nit: can be `and not self.has_access_to_all_contexts()`

**Review by long74100 — APPROVED** [review]

File: enterprise/models.py (hunk 1 above)
Comment (christopappas) [reply enterprise/models.py:2868]: I can add a log line for sure 👍 

For what it's worth, models' save() method already raises ValidationError if a field is not correct (like if you try to set a bool field to a string)

File: tests/test_models.py [hunk 3]
```diff
@@ -2348,25 +2348,6 @@ def test_get_assignments_many_assignments(self):
...
-        """
-        user_with_no_link = factories.UserFactory(email='zelda@example.com')
-        SystemWideEnterpriseUserRoleAssignment.objects.get_or_create(
-            user=user_with_no_link,
-            role=roles_api.learner_role(),
-            enterprise_customer=None,
-            applies_to_all_contexts=False,
-        )
-
-        self.assertEqual(
```
Comment (christopappas) [inline tests/test_models.py:2365]: making a note here that i did examine https://github.com/openedx/edx-enterprise/blob/master/enterprise/management/commands/update_role_assignments_with_customers.py and it appears that we no longer create role assignments without either an enterprise_customer or applies_to_all_contexts; therefore this test should be safe to remove (without losing functional coverage)

**Pull request merged**